diff --git a/group.install b/group.install
index 19d6442..3895d37 100644
--- a/group.install
+++ b/group.install
@@ -429,62 +429,12 @@ function group_update_8014() {
  * Update the group members view to use the new cache group permissions context.
  */
 function group_update_8015() {
-  if (\Drupal::moduleHandler()->moduleExists('views')) {
-    $view = \Drupal::configFactory()->getEditable('views.view.group_members');
-    if (!$view->isNew()) {
-      foreach (['default', 'page_1'] as $display) {
-        $cache_metadata_key = "display.$display.cache_metadata";
-        $cache_metadata = $view->get($cache_metadata_key);
-
-        // Remove the deleted cache context and add the updated metadata.
-        $key = array_search('group_membership.roles.permissions', $cache_metadata['contexts'], TRUE);
-        if ($key !== FALSE) {
-          unset($cache_metadata['contexts'][$key]);
-        }
-        $cache_metadata['max-age'] = -1;
-        $cache_metadata['contexts'][] = 'route.group';
-        $cache_metadata['contexts'][] = 'user.group_permissions';
-
-        $view->set($cache_metadata_key, $cache_metadata);
-      }
-
-      $view->save(TRUE);
-    }
-  }
+  // Empty update. See group_post_update_view_cache_contexts();
 }
 
 /**
  * Re-run a fixed version of _8015 and gnode_update_8006 on all views.
  */
 function group_update_8016() {
-  if (\Drupal::moduleHandler()->moduleExists('views')) {
-    $config_factory = \Drupal::configFactory();
-
-    foreach ($config_factory->listAll('views.view.') as $view_config_name) {
-      $view = $config_factory->getEditable($view_config_name);
-
-      $needs_save = FALSE;
-      foreach ($view->get('display') as $display => $display_data) {
-        $cache_metadata_key = "display.$display.cache_metadata";
-        $cache_metadata = $view->get($cache_metadata_key);
-
-        // Remove the deleted cache context and add the updated metadata.
-        $key = array_search('group_membership.roles.permissions', $cache_metadata['contexts'], TRUE);
-        if ($key !== FALSE) {
-          unset($cache_metadata['contexts'][$key]);
-
-          $cache_metadata['max-age'] = -1;
-          $cache_metadata['contexts'][] = 'route.group';
-          $cache_metadata['contexts'][] = 'user.group_permissions';
-
-          $view->set($cache_metadata_key, $cache_metadata);
-          $needs_save = TRUE;
-        }
-      }
-
-      if ($needs_save) {
-        $view->save(TRUE);
-      }
-    }
-  }
+  // Empty update. See group_post_update_view_cache_contexts();
 }
diff --git a/group.module b/group.module
index fe030ad..ea25b41 100644
--- a/group.module
+++ b/group.module
@@ -6,6 +6,7 @@
  */
 
 use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Cache\CacheBackendInterface;
 use Drupal\Core\Entity\ContentEntityInterface;
 use Drupal\Core\Entity\EntityInterface;
 use Drupal\Core\Field\FieldDefinitionInterface;
@@ -16,6 +17,7 @@ use Drupal\Core\Routing\RouteMatchInterface;
 use Drupal\Core\Session\AccountInterface;
 use Drupal\group\Entity\GroupContent;
 use Drupal\user\RoleInterface;
+use Drupal\views\ViewEntityInterface;
 
 /**
  * Implements hook_help().
@@ -573,3 +575,35 @@ function _group_content_enabler_manager() {
 function _group_role_synchronizer() {
   return \Drupal::service('group_role.synchronizer');
 }
+
+/**
+ * Implements hook_ENTITY_TYPE_presave().
+ *
+ * @todo Remove in 8.2.x if it still only fixes cache contexts by then.
+ */
+function group_view_presave(EntityInterface $view) {
+  $displays = $view->get('display');
+  $update = FALSE;
+
+  // Loop over all displays and replace the group_membership.roles.permissions
+  // cache context with its safe user.group_permissions counterpart.
+  foreach ($displays as $display_name => &$display) {
+    $cache_metadata = &$display['cache_metadata'];
+    if (isset($cache_metadata['contexts'])) {
+      // Remove the deleted cache context and add the updated metadata.
+      $key = array_search('group_membership.roles.permissions', $cache_metadata['contexts'], TRUE);
+      if ($key !== FALSE) {
+        $update = TRUE;
+        unset($cache_metadata['contexts'][$key]);
+
+        $cache_metadata['max-age'] = CacheBackendInterface::CACHE_PERMANENT;
+        $cache_metadata['contexts'][] = 'route.group';
+        $cache_metadata['contexts'][] = 'user.group_permissions';
+      }
+    }
+  }
+
+  if ($update) {
+    $view->set('display', $displays);
+  }
+}
diff --git a/group.post_update.php b/group.post_update.php
index 1b2df17..fef743c 100644
--- a/group.post_update.php
+++ b/group.post_update.php
@@ -5,6 +5,7 @@
  * Post update functions for Group.
  */
 
+use Drupal\Core\Config\Entity\ConfigEntityUpdater;
 use Drupal\group\Entity\GroupType;
 use Drupal\group\Entity\GroupContentType;
 use Drupal\user\Entity\Role;
@@ -45,3 +46,16 @@ function group_post_update_grant_access_overview_permission() {
     }
   }
 }
+
+/**
+ * Fix cache contexts in views.
+ */
+function group_post_update_view_cache_contexts(&$sandbox) {
+  if (!\Drupal::moduleHandler()->moduleExists('views')) {
+    return;
+  }
+  // This will trigger the catch-all fix in group_view_presave().
+  \Drupal::classResolver(ConfigEntityUpdater::class)->update($sandbox, 'view', function ($view) {
+    return TRUE;
+  });
+}
diff --git a/modules/gnode/gnode.install b/modules/gnode/gnode.install
index 1e85757..7f709d1 100644
--- a/modules/gnode/gnode.install
+++ b/modules/gnode/gnode.install
@@ -122,26 +122,5 @@ function gnode_update_8005() {
  * Update the group nodes view to use the new cache group permissions context.
  */
 function gnode_update_8006() {
-  if (\Drupal::moduleHandler()->moduleExists('views')) {
-    $view = \Drupal::configFactory()->getEditable('views.view.group_nodes');
-    if (!$view->isNew()) {
-      foreach (['default', 'page_1'] as $display) {
-        $cache_metadata_key = "display.$display.cache_metadata";
-        $cache_metadata = $view->get($cache_metadata_key);
-
-        // Remove the deleted cache context and add the updated metadata.
-        $key = array_search('group_membership.roles.permissions', $cache_metadata['contexts'], TRUE);
-        if ($key !== FALSE) {
-          unset($cache_metadata['contexts'][$key]);
-        }
-        $cache_metadata['max-age'] = -1;
-        $cache_metadata['contexts'][] = 'route.group';
-        $cache_metadata['contexts'][] = 'user.group_permissions';
-
-        $view->set($cache_metadata_key, $cache_metadata);
-      }
-
-      $view->save(TRUE);
-    }
-  }
+  // Empty update. See group_post_update_view_cache_contexts();
 }
